controllers/gcpmachine_controller_test.go

```diff
@@ -48,4 +64,171 @@ var _ = Describe("GCPMachineReconciler", func() {
+			mockCtrl     *gomock.Controller
+			gcpService   *mocks.MockServiceInterface
+		)
+		BeforeEach(func() {
+			flag.Set("logtostderr", "false")
+			flag.Set("v", "2")
+			klog.SetOutput(GinkgoWriter)
+			var err error
+			bootstrapData := "data"
+			version := "v1.16.2"
+
+			mockCtrl = gomock.NewController(GinkgoT())
+			gcpService = mocks.NewMockServiceInterface(mockCtrl)
+
+			reconciler = &GCPMachineReconciler{
+				Client:     k8sClient,
+				Log:        log.Log,
+				computeSvc: gcpService,
+			}
+
+			gcpMachine := &infrav1.GCPMachine{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "foo",
+					Namespace: "default",
+				},
+				Spec: infrav1.GCPMachineSpec{
+					Zone:         "us-west4-a",
+					InstanceType: "n1-standard-1",
+				},
+			}
+
+			machine := &clusterv1.Machine{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "fooMachine",
+					Namespace: "default",
+				},
+				Spec: clusterv1.MachineSpec{
+					InfrastructureRef: corev1.ObjectReference{
+						APIVersion: "infrastructure.cluster.x-k8s.io/v1alpha3",
+						Kind:       "GCPMachine",
+						Name:       gcpMachine.Name,
+					},
+					Bootstrap: clusterv1.Bootstrap{
+						Data: &bootstrapData,
+					},
+					Version: &version,
+				},
+			}
+
+			Expect(k8sClient.Create(ctx, gcpMachine)).Should(Succeed())
+
+			cluster := &clusterv1.Cluster{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "cluster",
+					Namespace: "default",
+				},
+				Status: clusterv1.ClusterStatus{
+					InfrastructureReady: true,
+				},
+			}
+
+			gcpCluster := &infrav1.GCPCluster{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "gcpcluster",
+					Namespace: "default",
+				},
+				Spec: infrav1.GCPClusterSpec{
+					Project: "gcp-project",
+				},
+			}
+
+			// Create the cluster scope
+			clusterScope, err = scope.NewClusterScope(scope.ClusterScopeParams{
+				Client:     fake.NewFakeClient(),
+				Logger:     log.Log,
+				Cluster:    cluster,
+				GCPCluster: gcpCluster,
+			})
+			Expect(err).To(BeNil())
+
+			// Create the machine scope
+			machineScope, err = scope.NewMachineScope(scope.MachineScopeParams{
+				Logger:     log.Log,
+				Client:     fake.NewFakeClient(),
+				Cluster:    cluster,
+				Machine:    machine,
+				GCPCluster: gcpCluster,
+				GCPMachine: gcpMachine,
+			})
+			Expect(err).To(BeNil())
+
+		})
+
+		AfterEach(func() {
+			Expect(k8sClient.Delete(ctx, machineScope.GCPMachine)).Should(Succeed())
+		})
+
+		It("should return early if gcp machine is in error state", func() {
+			machineScope.SetErrorReason(capierrors.UpdateMachineError)
+			machineScope.SetErrorMessage(errors.New("GCE error"))
+			Expect(k8sClient.Update(ctx, machineScope.GCPMachine)).Should(Succeed())
+
+			buf := new(bytes.Buffer)
+			klog.SetOutput(buf)
+
+			_, err := reconciler.reconcile(ctx, machineScope, clusterScope)
+			Expect(err).To(BeNil())
+			Expect(buf.String()).To(ContainSubstring("Error state detected, skipping reconciliation"))
```

Comment:
Checking for a specific log message seems a bit brittle. Do we need a better way to bubble this information up?

Reply:
The controller currently sets the machineScope.info with this message and intentionally sets the error to nil, would you prefer returning a specific error?

Reply:
I believe there was an issue with returning an error and having a way to tell controller-runtime not to requeue the request. According to https://github.com/kubernetes-sigs/controller-runtime/issues/617 it looks like this is still an outstanding issue.

Reply:
there is a series of checks that happen before this call is made, so I don't think just checking for `InstanceIfExists` not being called would be enough replacement that the right check failed